### DISCUSSION TITLE
I2S Audio on M5 StickCPlus with PAM8303 Speaker Hat - How to compile & configure a tasmota32-lvgl binary

### DISCUSSION THREAD

**Discussion by rudolphchg** [arendst/Tasmota#19097] [Q&A]
I'd be very grateful for any help concerning this Tasmota32 (Qasim) problem:

I am unable to hear sound from this speaker hat (PAM8303) attached to the M5 StickCPlus after loading the 
self-compiled binary (master) on gitpod:           
using "patformio run -e tasmota32-lvgl" 
with this addtl setting in user_config_override.h and the inclusion of the two necessary audio libraries in lib_extra.
I use the internal DAC for the StickCPlus (USE_I2S_NO_DAC)  as I only have the speaker hat which is just an amplifier.

#ifndef USE_I2S_AUDIO
#define USE_I2S_AUDIO                             // Add support for I2S audio output
#define USE_I2S_NO_DAC                         // Add support for transistor-based output without DAC
//#define USE_I2S_LSB                            // Add support for LSBJ chips, e.g. TM8211/PT8211      -> makes no difference 
#define USE_I2S_WEBRADIO                       // Add support for MP3 web radio streaming (only on ESP32 with PSRAM)
#define USE_I2S_SAY_TIME                       // Add support for english speaking clock
#define USE_I2S_RTTTL                          // Add support for Rtttl playback
#endif

Below is the complete output on the console after restart:

00:00:00.002 HDW: ESP32-PICO-D4 
00:00:00.049 UFS: FlashFS mounted with 296 kB free
00:00:00.066 CFG: Loaded from File, Count 373
00:00:00.074 QPC: Count 1
00:00:00.076 CFG: CR 432/699, Busy 0
00:00:00.079 SPI: Hardware using GPIO13(CLK) and GPIO15(MOSI)
00:00:00.081 I2C: Bus1 using GPIO33(SCL) and GPIO32(SDA)
00:00:00.082 I2C: Bus2 using GPIO22(SCL) and GPIO21(SDA)
00:00:00.094 I2C: BM8563 found at 0x51 (bus2)
00:00:00.096 RTC: UTC 2023-07-11T19:13:26, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
20:13:26.001 RTC: Synced by BM8563
20:13:26.007 TFS: File 'mcp23x.dat' not found
20:13:26.008 MCP: Invalid template
20:13:26.011 ROT: Mode 1
20:13:26.152 BRY: Berry initialized, RAM used=5670 bytes
20:13:26.183 CFG: loading M5StickC.autoconf#preinit.be
20:13:26.383 I2C: AXP192 detected on bus 2
20:13:26.401 CFG: loaded  M5StickC.autoconf#preinit.be
20:13:26.411 BRY: No 'preinit.be'
20:13:26.415 TFS: File '.drvset003' not found
20:13:26.416 CFG: Energy use defaults as file system not ready or file not found
20:13:26.444 DSP: File descriptor used
20:13:26.444 DSP: Configured display 'ST7789'
20:13:26.452 SRC: Restart
20:13:26.461 Project tasmota - Tasmota Version 13.0.0(lvgl-haspmota)-2_0_10(2023-07-11T18:34:47)
20:13:26.470 LVG: LVGL initialized
20:13:26.545 Init audio I2S: port=0, bclk=16, ws=-1, dout=26, mclk=17, din=-1   <============== !!!!
20:13:26.547 ETH: No ETH MDC and/or ETH MDIO GPIO defined
20:13:26.699 I2C: MPU6886 found at 0x68 (bus2)
20:13:26.948 BRY: No 'autoexec.be'
20:13:27.207 WIF: Attempting connection...
20:13:27.650 WIF: Connecting to AP1 Asteroid Channel 9 BSSId 10:FE:ED:7B:67:62 in mode 11n as tasmota-00A3BC-0956...
20:13:28.047 WIF: IPv4 192.168.3.103, mask 255.255.255.0, gateway 192.168.3.1
20:13:28.136 WIF: Connected
20:13:28.396 HTP: Web server active on tasmota-00A3BC-0956 with IP address 192.168.3.103
20:13:29.028 RTC: UTC 2023-07-11T19:13:31, DST 2023-03-26T02:00:00, STD 2023-10-29T03:00:00
20:13:31.000 RTC: Synced by NTP
20:13:31.478 WIF: IPv6 Local fe80::ea9f:6dff:fe00:a3bc
20:13:31.683 RSL: INFO1 = {"Info1":{"Module":"M5StickC","Version":"13.0.0(lvgl-haspmota)","FallbackTopic":"cmnd/DVES_00A3BC_fb/","GroupTopic":"cmnd/tasmotas/"}}
20:13:31.697 RSL: INFO2 = {"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-00A3BC-0956","IPAddress":"192.168.3.103","IP6Global":"","IP6Local":"fe80::ea9f:6dff:fe00:a3bc"}}
20:13:31.710 RSL: INFO3 = {"Info3":{"RestartReason":"Software reset CPU","BootCount":97}}
20:13:31.713 RSL: RESULT = {"POWER":"ON"}
20:13:31.724 RSL: POWER = ON
20:13:31.729 TFS: File 'autoexec.bat' not found
20:13:35.580 HTP: Main Menu
20:13:35.977 QPC: Reset
20:13:35.985 RSL: STATE = {"Time":"2023-07-11T20:13:35","Uptime":"0T00:00:09","UptimeSec":9,"Heap":27,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":99,"MqttCount":0,"Berry":{"HeapUsed":7,"Objects":90},"POWER":"ON","Wifi":{"AP":1,"SSId":"Asteroid","BSSId":"10:FE:ED:7B:67:62","Channel":9,"Mode":"11n","RSSI":54,"Signal":-73,"LinkCount":1,"Downtime":"0T00:00:02"}}
20:13:36.025 RSL: SENSOR = {"Time":"2023-07-11T20:13:36","MPU6886":{"AX":135,"AY":121,"AZ":1049,"GX":0,"GY":9,"GZ":-7},"AXP192":{"VBusVoltage":5.058,"VBusCurrent":5.0,"BattVoltage":4.145,"BattCurrent":81.5,"Temperature":63.3},"TempUnit":"C"}
20:13:37.958 APP: Boot Count 97
20:13:38.415 CFG: Saved, Count 374, Bytes 4096
20:13:47.519 HTP: Consoles
20:13:49.874 HTP: Console

Here, the template has been configured for GPIO26 as I2S_DOUT (Channel 1), and BCLK and MCLK have been set to GPIO 16 and 17.
However, I have also tried all combinations of leaving BCLK and MCLK set to None or User, as well as variations
using I2S_DOUT set to GPIO 25 or 26.
As checked in xdrvr_42, the port used is always 0 (corresponding to Channel 1). SI observe that specifying only 
I2S_DOUT on Channel 2 prevents any output of Init Audio I2S and does not allow changes of  I2SGain values.

Here are my console commands which let me believe that the software compilation is ok, but that my configuration
is still buggy, despite all the attempts, as I cannot hear any changes in the sound when launching the commands:

20:14:06.978 CMD: I2SGain 100
20:14:06.979 SRC: WebConsole from 192.168.3.100
20:14:06.980 CMD: Grp 0, Cmd 'I2SGAIN', Idx 1, Len 3, Pld 100, Data '100'
20:14:06.983 RSL: RESULT = {"I2SGain":100}
20:14:25.510 CMD: I2SSay Hello Hello
20:14:25.511 SRC: WebConsole from 192.168.3.100
20:14:25.513 CMD: Grp 0, Cmd 'I2SSAY', Idx 1, Len 11, Pld -99, Data 'Hello Hello'
20:14:26.450 RSL: RESULT = {"I2SSay":"Hello Hello"}

The speaker hat is ok, I have used it lately with ESPRadioMP3.
I have tried understanding the settings in ESPRadioMP3, but I am too much of a
beginner that I cannot find the correct ones to transfer them to Tasmota....
(At least LVGL works nicely after having overcome the hurdles with display.ini and the DisplayMode=0 setting.)

So in case, there is a kind soul out there who has already successfully  tried Tasmota on a StickCplus, please let me know.

**Comment by s-hadinger** [comment]
M5stickc requires that the axp192 sends power to the 5V pin on the Grove connector. It's normally done by the driver here 

https://github.com/arendst/Tasmota/blob/889e6ea54d8541ba713eb305d7d15d7695abc5e5/tasmota/berry/drivers/AXP192_M5StickC.be#L45

Can you check that you get 5V on the Grove connector?

**Reply by s-hadinger** [reply]
Actually it's done in autoconf here 
https://github.com/tasmota/autoconf/blob/df3e8cbb71046814e2f37fbe5998053e925326bd/raw/esp32/M5StickC/preinit.be#L45

**Comment by rudolphchg** [comment]
Thanks very much for your swift reply (just one hour after my entry!)
I checked it on the Hat Connector ( ref pictures in pic.zip ) which is
actually the relevant one.
I cannot check the grove connector since I do not own a corresponding
connector yet,
Normal Dupont cables are too large for this tiny connector to fit.

So, there is indeed 5V power on the Hat, but no sound apart from a veeery
faint crackling noise.
Anychanges in sound just correlate with the internal LED.

Here again the current settings (as of the photo) in the console:

11:35:25.556 LVG: LVGL initialized
11:35:25.631 Init audio I2S: port=0, bclk=-1, ws=-1, dout=26, mclk=-1,
din=34
11:35:25.790 I2C: MPU6886 found at 0x68 (bus2)
11:35:26.816 WIF: Connecting to AP1 Asteroid Channel 9 BSSId
10:FE:ED:7B:67:62 in mode 11n as tasmota-00A3BC-0956...
11:35:27.125 WIF: Connected
11:35:27.391 HTP: Web server active on tasmota-00A3BC-0956 with IP address
192.168.3.109
11:35:31.700 RSL: INFO1 =
{"Info1":{"Module":"M5StickCPlus","Version":"13.0.0(lvgl-haspmota)","FallbackTopic":"cmnd/DVES_00A3BC_fb/","GroupTopic":"cmnd/tasmotas/"}}
11:35:31.715 RSL: INFO2 =
{"Info2":{"WebServerMode":"Admin","Hostname":"tasmota-00A3BC-0956","IPAddress":"192.168.3.109","IP6Global":"","IP6Local":"fe80::ea9f:6dff:fe00:a3bc"}}
11:35:31.727 RSL: INFO3 = {"Info3":{"RestartReason":"Vbat power on
reset","BootCount":38}}
11:35:31.730 RSL: RESULT = {"POWER":"ON"}
11:35:31.741 RSL: POWER = ON
11:35:36.130 QPC: Reset
11:35:36.136 RSL: STATE =
{"Time":"2023-07-13T11:35:36","Uptime":"0T00:00:11","UptimeSec":11,"Heap":26,"SleepMode":"Dynamic","Sleep":10,"LoadAvg":91,"MqttCount":0,"Berry":{"HeapUsed":7,"Objects":90},"POWER":"ON","Wifi":{"AP":1,"SSId":"Asteroid","BSSId":"10:FE:ED:7B:67:62","Channel":9,"Mode":"11n","RSSI":70,"Signal":-65,"LinkCount":1,"Downtime":"0T00:00:02"}}
11:35:36.176 RSL: SENSOR =
{"Time":"2023-07-13T11:35:36","MPU6886":{"AX":-536,"AY":87,"AZ":917,"GX":-2,"GY":8,"GZ":-6},"AXP192":{"VBusVoltage":0.687,"VBusCurrent":0.7,"BattVoltage":4.037,"BattCurrent":-94.5,"Temperature":46.7},"TempUnit":"C"}
11:37:26.701 CMD: I2SGain 100
11:37:26.705 RSL: RESULT = {"I2SGain":100}
11:37:38.681 CMD: I2SSay Hello
11:37:39.155 RSL: RESULT = {"I2SSay":"Hello"}

Kind regards,
Christine

s-hadinger ***@***.***> schrieb am Di., 11. Juli 2023, 22:26:

**Comment by nikvostro** [comment]
@rudolphchg hi!
I'm also trying to run my ESP32 with USE_I2S_NO_DAC but with no luck... tried different combinations of USE_I2S_... definitions and assiging GPIO25 for ESP32-WROOM and GPIO17 for ESP32-S as I2S DOUT pin. FW compilation completes fine every time. Tested those GPIOs as PWM - they are working well (not broken). 
Nothing comes out on the specified pins (checking via oscilloscope) on execution of any kind of i2s command like i2splay (wav/mp3), i2ssay, i2stime etc...

I'm adopting this project for my needs and want to get sound generation working https://github.com/SCratORS/SmartIntercom. For this project with author's FW sound works well through GPIO25 pin being connected to internal DAC. Looking for some help on [Tasmota's Discord](https://discord.com/channels/479389167382691863/790917300210368562/1130762833341202462) - but nobody came yet.

Do you have any success so far? Apart of this case maybe.
